File: src/main/java/com/filemanager/app/MainActivity.java
package com.filemanager.app;

import android.app.Activity;
import android.os.Bundle;
import android.os.Environment;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;

public class MainActivity extends Activity {

    private ListView fileListView;
    private TextView pathTextView;
    private ArrayAdapter<String> adapter;
    private ArrayList<String> fileList;
    private File currentDirectory;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        fileListView = findViewById(R.id.fileListView);
        pathTextView = findViewById(R.id.pathTextView);
        Button backButton = findViewById(R.id.backButton);
        Button homeButton = findViewById(R.id.homeButton);

        fileList = new ArrayList<>();
        adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, fileList);
        fileListView.setAdapter(adapter);

        currentDirectory = Environment.getExternalStorageDirectory();
        loadFiles();

        backButton.setOnClickListener(v -> goBack());
        homeButton.setOnClickListener(v -> goHome());

        fileListView.setOnItemClickListener((parent, view, position, id) -> {
            String selectedFile = fileList.get(position);
            File selected = new File(currentDirectory, selectedFile);
            if (selected.isDirectory()) {
                currentDirectory = selected;
                loadFiles();
            } else {
                Toast.makeText(MainActivity.this, "File: " + selectedFile, Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void loadFiles() {
        fileList.clear();
        pathTextView.setText("📂 " + currentDirectory.getAbsolutePath());

        File[] files = currentDirectory.listFiles();
        if (files != null) {
            ArrayList<String> dirs = new ArrayList<>();
            ArrayList<String> fileNames = new ArrayList<>();

            for (File file : files) {
                if (file.getName().startsWith(".")) continue;
                if (file.isDirectory()) {
                    dirs.add("📁 " + file.getName());
                } else {
                    fileNames.add("📄 " + file.getName());
                }
            }

            Arrays.sort(dirs.toArray(new String[0]));
            Arrays.sort(fileNames.toArray(new String[0]));

            fileList.addAll(dirs);
            fileList.addAll(fileNames);
        }

        adapter.notifyDataSetChanged();
    }

    private void goBack() {
        File parent = currentDirectory.getParentFile();
        if (parent != null) {
            currentDirectory = parent;
            loadFiles();
        } else {
            Toast.makeText(this, "Already at root", Toast.LENGTH_SHORT).show();
        }
    }

    private void goHome() {
        currentDirectory = Environment.getExternalStorageDirectory();
        loadFiles();
    }
}

File: app/src/main/java/com/example/filemanager/MainActivity.kt
package com.example.filemanager

import android.net.Uri
import android.os.Bundle
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.ListView
import android.widget.TextView
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.documentfile.provider.DocumentFile

class MainActivity : AppCompatActivity() {
    private var currentUri: Uri? = null
    private var navigationStack = mutableListOf<Uri>()
    private lateinit var fileListView: ListView
    private lateinit var pathTextView: TextView
    private lateinit var backButton: Button
    private lateinit var selectButton: Button
    private var fileItems = mutableListOf<FileItem>()
    private lateinit var adapter: ArrayAdapter<String>

    private val selectFolderLauncher = registerForActivityResult(
        ActivityResultContracts.OpenDocumentTree()
    ) { uri: Uri? ->
        if (uri != null) {
            currentUri = uri
            navigationStack.clear()
            navigationStack.add(uri)
            loadFiles(uri)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        fileListView = findViewById(R.id.fileListView)
        pathTextView = findViewById(R.id.pathTextView)
        backButton = findViewById(R.id.backButton)
        selectButton = findViewById(R.id.selectButton)

        adapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, mutableListOf<String>())
        fileListView.adapter = adapter

        selectButton.setOnClickListener {
            selectFolderLauncher.launch(null)
        }

        backButton.setOnClickListener {
            if (navigationStack.size > 1) {
                navigationStack.removeAt(navigationStack.size - 1)
                loadFiles(navigationStack.last())
            }
        }

        fileListView.setOnItemClickListener { _, _, position, _ ->
            if (position < fileItems.size) {
                val item = fileItems[position]
                if (item.isDirectory && item.uri != null) {
                    navigationStack.add(item.uri)
                    loadFiles(item.uri)
                }
            }
        }

        pathTextView.text = "No folder selected"
        backButton.isEnabled = false
    }

    private fun loadFiles(folderUri: Uri) {
        val documentFile = DocumentFile.fromTreeUri(this, folderUri) ?: return
        fileItems.clear()

        try {
            documentFile.listFiles().forEach { file ->
                fileItems.add(
                    FileItem(
                        name = file.name ?: "Unknown",
                        size = if (file.isFile) file.length() else 0,
                        isDirectory = file.isDirectory,
                        uri = file.uri,
                        lastModified = file.lastModified()
                    )
                )
            }
        } catch (e: Exception) {
            // Handle permission errors
        }

        // Sort: directories first, then by name
        fileItems.sortWith(compareBy({ !it.isDirectory }, { it.name }))

        // Update UI
        val displayList = fileItems.map { item ->
            val type = if (item.isDirectory) "[DIR]" else "[FILE]"
            val sizeStr = if (item.isDirectory) "" else " (${formatSize(item.size)})"
            "${item.name} $type$sizeStr"
        }

        adapter.clear()
        adapter.addAll(displayList)
        adapter.notifyDataSetChanged()

        pathTextView.text = documentFile.name ?: "File Manager"
        backButton.isEnabled = navigationStack.size > 1
    }

    private fun formatSize(bytes: Long): String {
        return when {
            bytes >= 1024 * 1024 * 1024 -> String.format("%.2f GB", bytes / (1024f * 1024f * 1024f))
            bytes >= 1024 * 1024 -> String.format("%.2f MB", bytes / (1024f * 1024f))
            bytes >= 1024 -> String.format("%.2f KB", bytes / 1024f)
            else -> "$bytes B"
        }
    }

    data class FileItem(
        val name: String,
        val size: Long,
        val isDirectory: Boolean,
        val uri: Uri?,
        val lastModified: Long
    )
}
